zrml/court/src/tests.rs

```diff
@@ -51,297 +66,2342 @@ const DEFAULT_MARKET: MarketOf<Runtime> = Market {
+        let outcome = OutcomeReport::Scalar(42u128);
+        let (market_id, _, salt) = set_alice_after_vote(outcome.clone());
+
+        run_blocks(CourtVotePeriod::get() + 1);
+
+        assert_ok!(Court::reveal_vote(Origin::signed(ALICE), market_id, outcome.clone(), salt));
+
+        assert_noop!(
+            Court::reveal_vote(Origin::signed(ALICE), market_id, outcome, salt),
+            Error::<Runtime>::VoteAlreadyRevealed
+        );
+    });
+}
+
+#[test]
+fn reveal_vote_fails_if_already_denounced() {
+    ExtBuilder::default().build().execute_with(|| {
+        let outcome = OutcomeReport::Scalar(42u128);
+        let (market_id, _, salt) = set_alice_after_vote(outcome.clone());
+
+        assert_ok!(Court::denounce_vote(
+            Origin::signed(BOB),
+            market_id,
+            ALICE,
+            outcome.clone(),
+            salt
+        ));
+
+        run_blocks(CourtVotePeriod::get() + 1);
+
+        assert_noop!(
+            Court::reveal_vote(Origin::signed(ALICE), market_id, outcome, salt),
+            Error::<Runtime>::VoteAlreadyDenounced
+        );
+    });
+}
+
+#[test]
+fn denounce_vote_works() {
+    ExtBuilder::default().build().execute_with(|| {
+        let outcome = OutcomeReport::Scalar(42u128);
+        let (market_id, secret, salt) = set_alice_after_vote(outcome.clone());
+
+        let old_draws = <Draws<Runtime>>::get(market_id);
+        assert!(
+            old_draws
+                .iter()
+                .any(|draw| { draw.juror == ALICE && matches!(draw.vote, Vote::Secret { .. }) })
+        );
+
+        let free_alice_before = Balances::free_balance(ALICE);
+        let pot_balance_before = Balances::free_balance(&Court::reward_pot(&market_id));
+
+        assert_ok!(Court::denounce_vote(
+            Origin::signed(BOB),
+            market_id,
+            ALICE,
+            outcome.clone(),
+            salt
+        ));
+        System::assert_last_event(
+            Event::DenouncedJurorVote {
+                denouncer: BOB,
+                juror: ALICE,
+                market_id,
+                outcome: outcome.clone(),
+                salt,
+            }
+            .into(),
+        );
+
+        let new_draws = <Draws<Runtime>>::get(market_id);
+        assert_eq!(old_draws[1..], new_draws[1..]);
+        assert_eq!(old_draws[0].juror, ALICE);
+        assert_eq!(old_draws[0].juror, new_draws[0].juror);
+        assert_eq!(old_draws[0].weight, new_draws[0].weight);
+        assert_eq!(old_draws[0].slashable, new_draws[0].slashable);
+        assert_eq!(old_draws[0].vote, Vote::Secret { secret });
+        assert_eq!(new_draws[0].vote, Vote::Denounced { secret, outcome, salt });
+
+        <JurorPool<Runtime>>::get().iter().for_each(|pool_item| {
+            assert_ne!(pool_item.juror, ALICE);
+        });
+
+        let free_alice_after = Balances::free_balance(ALICE);
+        let slash = old_draws[0].slashable;
+        assert!(!slash.is_zero());
```

Comment:
Good thought! `denounce` is kind of a special case in the court system, because it is a slash that is not handled by `reassign_juror_stakes`. That's what I have changed now. Since in `reassign_juror_stakes` the market inside `Draws` is removed. So the `slashable` item is non-existent at that point anyways.

49f05cd66edc5f84e9812c80fd4155a644b5ce1f

Reply:
The juror `stake` is only decreased [here](https://github.com/zeitgeistpm/zeitgeist/blob/aced219db47124e4f907bc51d7f201be87f3684a/zrml/court/src/lib.rs#L518). This happens if the juror decided to call `prepare_exit_court`. This juror is not able to get selected for future court rounds. The `stake` shouldn't be changed elsewhere, because it's part of the key of the `binary_search_by_key` function.